### PR TITLE
Proposing slide content for Scenario 2 - resolvers describing themselves

### PR DESCRIPTION
**Pull request by tojens-ietf**

### CODE REVIEW

File: interim-jan-2021/resolver-behavior-self-description.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Resolver behavior self-description
+
+## Defining local-only namespaces
```
Comment (bemasc) [inline interim-jan-2021/resolver-behavior-self-description.md:3]: I think we ought to have a requirement that the protocol enables clients to check that the local namespace is authorized, if it collides with a global namespace.

File: interim-jan-2021/resolver-behavior-self-description.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+* Express namespaces which only this resolver can resolve
+* Ex 1: Enterprise resolvers serving corporation-specific namespaces
+* Ex 2: Public Wi-Fi or cellular networks offering network-local services
+
+## Defining per-namespace optimization
+* Express namespaces for which this resolver provides preferable resolutions
+* Ex 1: ISP routes public content requests to network caches
+* Ex 2: Cellular provider charges less for content from specific servers
```
Comment (bemasc) [inline interim-jan-2021/resolver-behavior-self-description.md]: Is this a thing?  I've never heard of DNS being used this way.

File: interim-jan-2021/resolver-behavior-self-description.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# Resolver behavior self-description
+
+## Defining local-only namespaces
+* Express namespaces which only this resolver can resolve
+* Ex 1: Enterprise resolvers serving corporation-specific namespaces
+* Ex 2: Public Wi-Fi or cellular networks offering network-local services
+
+## Defining per-namespace optimization
+* Express namespaces for which this resolver provides preferable resolutions
+* Ex 1: ISP routes public content requests to network caches
```
Comment (bemasc) [inline interim-jan-2021/resolver-behavior-self-description.md:11]: Is this really defined by a namespace?  I don't suppose you're going to enumerate all of Cloudflare's domains...

File: interim-jan-2021/resolver-behavior-self-description.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+* Ex 2: Cellular provider charges less for content from specific servers
+* Ex 3: Public resolver designated to serve a namespace to limit parties privy to resolution
+
+## Defining answer modification behavior
+* Express how query answers are modified by the resolver (but not necessarily _why_)
+* Ex 1: Will not resolve names breaching resolver policy
```
Comment (bemasc) [inline interim-jan-2021/resolver-behavior-self-description.md]: Surely no resolver will resolve names that breach its policy?  This seems tautological, so I don't think we need a way to express it.

File: interim-jan-2021/resolver-behavior-self-description.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+* Express namespaces for which this resolver provides preferable resolutions
+* Ex 1: ISP routes public content requests to network caches
+* Ex 2: Cellular provider charges less for content from specific servers
+* Ex 3: Public resolver designated to serve a namespace to limit parties privy to resolution
+
+## Defining answer modification behavior
+* Express how query answers are modified by the resolver (but not necessarily _why_)
+* Ex 1: Will not resolve names breaching resolver policy 
+    * May or may not define applicable namespaces
+* Ex 2: Will drop specific record types
```
Comment (bemasc) [inline interim-jan-2021/resolver-behavior-self-description.md]: What's the use case for this?  RFC 3597 is pretty clear that this is not allowed, and I've never encountered a use case where it was needed.

File: interim-jan-2021/resolver-behavior-self-description.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    * Helps client distinguish errors from by-design failures
+
+## Defining protocol support
+* Express what optional DNS-related functionality is supported 
+* Ex 1: DNS Extended Errors and which codes to expect
+* Ex 2: Which encryption protocols are supported
+    * DoH and DoT, DoT only, etc.
+    * Should be extensible to accommodate future protocols
+* Ex 3: Requires some form of client identity confirmation
```
Comment (bemasc) [inline interim-jan-2021/resolver-behavior-self-description.md]: There's something funny about this.  If the server is access-controlled, then haven't you already authenticated before you can get this information?  Perhaps the requirement is that access-controlled servers have a way to make this information be not-access-controlled?

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 1 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md:3]: Good point. I'll address with an update.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 2 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md]: One concrete example of this is Netflix Open Connect, which allows ISPs to acquire hardware from Netflix to be hosted in their local data centers and then control routing from their customers so they can serve Netflix content from their own data centers.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 5 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md]: This isn't a scenario I'm happy to support, but I included it as there are bound to be servers who choose to do so anyway (say networks which forbid ECH and similar technologies). It would be useful for clients to know this is by design rather than see other issue they should investigate.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 4 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md]: My intended point is to allow resolvers to communicate whether they do any modification of answers at all. Some resolvers don't such as those qualifying for Mozilla's TRR program. Some block ad domains. Some don't. 

I think this is a scenario that has gathered sufficient interest in the WG to document as a requirement for drafts to meet. I think it's sufficiently different from policy (describes resolver behavior without mandating client behavior) to not be a non-goal. 

Open to discussion.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 6 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md]: Yes, your phrasing is much clearer. I'll update with next commit.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 2 above)
Comment (bemasc) [reply interim-jan-2021/resolver-behavior-self-description.md]: Sure, but I've never heard of a cellular provider that charges their customers less for content loaded from the local Netflix cache node than they do for content from the Netflix backend servers.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 4 above)
Comment (bemasc) [reply interim-jan-2021/resolver-behavior-self-description.md]: > My intended point is to allow resolvers to communicate whether they do any modification of answers at all.

I find this difficult to reason about because (a) it's a forward-looking statement and (b) even resolvers that don't normally reject any queries probably will do so under some exceptional circumstance (e.g. a DoS attack).

I think the EDE "which codes to expect" requirement is a better way to draw this distinction.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 5 above)
Comment (bemasc) [reply interim-jan-2021/resolver-behavior-self-description.md]: I don't think we should support this.  I'm not aware of any present need for it, and it directly contravenes correct DNS behavior.

> networks which forbid ECH and similar technologies

It is not logical for a resolver to do this based on RR type.  If the operator wants to prevent access to a domain, it can do so directly from the QNAME, rather than trying to force the client to leak the name again (which is not even guaranteed to work, due to connection coalescing).

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 2 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md]: Ah, I misread which line was being commented on and was responding to Ex 1. Ignore my previous comment.

I was describing cases where name-based policies are used to determine network interfaces for cellular devices so that some requests work regardless of billing plan state (messages, billing portal, etc.). Obviously this can be solved device-side as well. 

It's come up in vague details in prior EDDI/ADD discussions, but I can't defend it more specifically. Happy to remove and let the WG argue it in from actual cellular folks.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 4 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md]: It's fair to say that the EDE line could cover this use case with a much cleaner architecture. I'll remove this line.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 2 above)
Comment (bemasc) [reply interim-jan-2021/resolver-behavior-self-description.md]: It sounds like you're describing the case of a network (cellular plan that ran out of credit) that has access to a very limited subset of the internet.  I think in this case, the limitation is a property of the network, not the resolver, and asking a resolver about a network's behavior is out of scope.

In general, this sounds like a better fit for CAPPORT.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 5 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md]: Some other examples include resolvers only serving browsing needs (not resolving queries for email record types). While one can argue that a DNS resolver only resolving some QTYPEs is incorrect, [it certainly is supported in practice today](https://docs.microsoft.com/en-us/windows-server/networking/dns/deploy/apply-filters-on-dns-queries). If we're defining requirements for resolvers describing their own behavior, it makes sense to describe what queries they will not resolve in advance.

I suppose we could simply rely on EDE code 21 for this instead, but if that's our position, I would want to call that out in our non-goals so the path for this scenario is known.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 2 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md]: > I think in this case, the limitation is a property of the network, not the resolver

I agree. I'll remove this line.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 5 above)
Comment (bemasc) [reply interim-jan-2021/resolver-behavior-self-description.md]: I would prefer to make this a non-goal.

I note that the example given in that documentation is to refuse the "ANY" QTYPE, which is actually allowed (RFC 8482).  A flag for the "refuse-ANY" behavior could be interesting.  Perhaps a missing requirement here is extensibility, to allow adding more signals over time.  Then if the IETF someday decides that refusing MX queries is OK, we could add a flag for that.

EDIT: It looks like RFC 8482 is pretty clear about only applying to authoritative servers, not resolvers.

**Comment by bemasc** [comment]
I think this also needs something about conveying a URL for a webpage
describing this service.

**Comment by tojens-ietf** [comment]
See, I thought so too but left it out because I thought it was getting ahead of the actual requirements (we are assuming a web page is the best way to handle that). 

I'll add something to this effect and we can debate it. I do think it would be the best approach anyway. We should be careful in our description though: intended for human consumption, not machine legible, etc.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 2 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md]: I've issued a new commit with the discussed changes. 

Having said that, I just noticed CAPPORT has been concluded. This and other similar scenarios ADD is bringing to the surface (*cough* networks expressing DNS policy *cough*) will need a home that apparently does not exist.

File: interim-jan-2021/resolver-behavior-self-description.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+## Defining resolver identity
+* Express information consumable by humans describing the resolver's identity
+* Ex 1: Provide human-legible documentation
+    * Most likely a web page link to explain server identity, terms of use, etc.
+	* Not used for decision making by any protocol peer; communicated to clients for display to users
+
+## Defining protocol support
+* Express what optional DNS-related functionality is supported 
+* Ex 1: DNS Extended Errors and which codes to expect
```
Comment (bemasc) [inline interim-jan-2021/resolver-behavior-self-description.md:24]:
```suggestion
* Ex 1: DNS Extended Errors and which codes to expect
      * Not exhaustive: server can still return any code
      * Codes 15-17 indicate kinds of filtering the resolver implements
```

File: interim-jan-2021/resolver-behavior-self-description.md [hunk 8]
```diff
@@ -0,0 +1,28 @@
+# Resolver behavior self-description
+
+## Defining local-only namespaces
+* Express namespaces which only this resolver can resolve
+    * Authoritatively if the namespace collides with global namespaces
```
Comment (bemasc) [inline interim-jan-2021/resolver-behavior-self-description.md]:
```suggestion
    * Authoritatively if the namespace collides with any global names
```

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 3 above)
Comment (tireddy2) [reply interim-jan-2021/resolver-behavior-self-description.md:11]: Downstream CDNs are typically deployed in the ISP cloud for better user experience. For example, 5G discusses deploying downstream CDNs in the edge (e.g., video streaming content, gaming servers) . 

I don't think it is feasible to provide all the domains for which downstream CDN is deployed in the ISP cloud (edge or core).

File: interim-jan-2021/resolver-behavior-self-description.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+## Defining local-only namespaces
+* Express namespaces which only this resolver can resolve
+    * Authoritatively if the namespace collides with any global names
+* Ex 1: Enterprise resolvers serving corporation-specific namespaces
+* Ex 2: Public Wi-Fi or cellular networks offering network-local services
+
+## Defining per-namespace optimization
+* Express namespaces for which this resolver provides preferable resolutions
+* Ex 1: ISP routes public content requests to network caches
+* Ex 2: Public resolver designated to serve a namespace to limit parties privy to resolution
```
Comment (tireddy2) [inline interim-jan-2021/resolver-behavior-self-description.md:12]: I don't get this bullet. Are you referring to a public enterprise resolver serving the namespace to authenticated and authorized clients ?

File: interim-jan-2021/resolver-behavior-self-description.md [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+## Defining resolver identity
+* Express information consumable by humans describing the resolver's identity
+* Ex 1: Provide human-legible documentation
+    * Most likely a web page link to explain server identity, terms of use, etc.
+	* Not used for decision making by any protocol peer; communicated to clients for display to users
```
Comment (tireddy2) [inline interim-jan-2021/resolver-behavior-self-description.md:18]: Several WG members including you opposed any interaction with the end user. The argument was an average end user would not understand the privacy terms, server identity and encrypted DNS.

I started revising draft https://github.com/tireddy2/DNS-resolver-information/blob/master/draft-add-server-selection-information.txt accordingly :)

File: interim-jan-2021/resolver-behavior-self-description.md [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+## Defining protocol support
+* Express what optional DNS-related functionality is supported 
+* Ex 1: DNS Extended Errors and which codes to expect
+      * Not exhaustive: server can still return any code
+      * Codes 15-17 indicate kinds of filtering the resolver implements
+* Ex 2: Which encryption protocols are supported
+    * DoH and DoT, DoT only, etc.
+    * Should be extensible to accommodate future protocols
+* Ex 3: Extensible mechanism for indicating optional resolver behavior
```
Comment (tireddy2) [inline interim-jan-2021/resolver-behavior-self-description.md:28]: A compliant DNS proxy must not drop unknown RR types, see https://tools.ietf.org/html/bcp152#section-4.3.
Why would a non-complaint DNS proxy indicate non-support of the the BCP ?
The client would anyway know whether the server supports DNSSEC or not (what is the need for a explicit signal from the server) ?

File: interim-jan-2021/resolver-behavior-self-description.md [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+    * Supported record types, will it refuse QTYPE=ANY, etc.
```
Comment (tireddy2) [inline interim-jan-2021/resolver-behavior-self-description.md]: A compliant DNS proxy must not drop unknown RR types, see https://tools.ietf.org/html/bcp152#section-4.3.
Why would a non-complaint DNS proxy indicate non-support of the the BCP ?
The client would anyway know whether the server supports DNSSEC or not (what is the need for a explicit signal from the server) ?

File: interim-jan-2021/resolver-behavior-self-description.md [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+* Ex 4: Access-controlled resolvers describing their properties outside of access control
```
Comment (tireddy2) [inline interim-jan-2021/resolver-behavior-self-description.md]: I don't get the above line. Please elaborate.

File: interim-jan-2021/resolver-behavior-self-description.md [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+* Ex 1: Provide human-legible documentation
+    * Most likely a web page link to explain server identity, terms of use, etc.
+	* Not used for decision making by any protocol peer; communicated to clients for display to users
+
+## Defining protocol support
+* Express what optional DNS-related functionality is supported 
+* Ex 1: DNS Extended Errors and which codes to expect
+      * Not exhaustive: server can still return any code
+      * Codes 15-17 indicate kinds of filtering the resolver implements
+* Ex 2: Which encryption protocols are supported
```
Comment (tireddy2) [inline interim-jan-2021/resolver-behavior-self-description.md]: This is already covered in the existing requirements draft.

File: interim-jan-2021/resolver-behavior-self-description.md [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+* Express information consumable by humans describing the resolver's identity
+* Ex 1: Provide human-legible documentation
+    * Most likely a web page link to explain server identity, terms of use, etc.
+	* Not used for decision making by any protocol peer; communicated to clients for display to users
+
+## Defining protocol support
+* Express what optional DNS-related functionality is supported 
+* Ex 1: DNS Extended Errors and which codes to expect
+      * Not exhaustive: server can still return any code
+      * Codes 15-17 indicate kinds of filtering the resolver implements
```
Comment (tireddy2) [inline interim-jan-2021/resolver-behavior-self-description.md:26]: Unfortunately the error codes are not complete, the client does not know if the server is performing filtering to block malware or phishing.

Comment (bemasc) [reply interim-jan-2021/resolver-behavior-self-description.md:26]: I don't think it makes sense for the IETF to try to maintain a taxonomy of all of the reasons someone might want to block a domain.  That is a human question, not a technical question, and I think it belongs in the realm of human communication, e.g. on a webpage.  I don't want us to end up debating whether to add "blasphemy" to an IANA registry.  However, this list of requirements is not exhaustive, so it does not preclude the WG from defining such a feature, either now or later (via the extensibility required by point 3).

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 10 above)
Comment (bemasc) [reply interim-jan-2021/resolver-behavior-self-description.md:18]: From my perspective, the "designated resolver" upgrade scenario should be completely non-interactive.  I think these slides are about supporting other scenarios.  Some uses I can think of:
* Inspecting and debugging my current configuration
* Browsing a directory of public resolvers

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 12 above)
Comment (bemasc) [reply interim-jan-2021/resolver-behavior-self-description.md]: I think this is partly my fault.  I told Tommy that RFC 8482 allows DNS servers to refuse "QTYPE=ANY" queries ... but that only seems to apply to authoritative servers, not resolvers.  Apart from that, I agree that compliance requires "qtype neutrality".

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 10 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md:18]: > Several WG members including you opposed any interaction with the end user

Absolutely correct, and I still feel that way.

> The argument was an average end user would not understand the privacy terms, server identity and encrypted DNS.

And this will always continue to be true IMO. Please don't stop your draft direction change :) unless what I'm saying below clarifies my previous "get policy out of my upgrade process" verbiage.

I agree with Ben: this doesn't mean "make the upgrade process interactive" to me. Rather, this is a way to provide "other info about this resolver intended for humans" to be considered by an auditor or debugger. 

Example: if there was a standard way to return a documentation link, Google DoH servers could return [this page](https://developers.google.com/speed/public-dns/docs/doh/). It would be nice as a client implementor if there was a standard way to retrieve per-resolver documentation.

Basically, this is about returning man pages that might include terms of server use, not about returning EULAs which require an agreement to the terms of use. The pivot point is about it being a part of the encryption or upgrade negotiation or not. I am definitely opposed to this link being considered a passive agreement (or being made a requirement for active agreement) to terms of use on the server. 

Does this help?

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 13 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md]: This is saying "if an encrypted resolver supports client authentication, it should be able to describe the requirements to becoming authenticated" so that a naive client or a client with invalid credentials can understand what is happening.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 15 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md:26]: > the client does not know if the server is performing filtering to block malware or phishing.

And I think that's ok. Such details are not actionable by the DNS client anyway. For example:
- Error 15, Blocked: it's the resolver's choice (the error could be bypassed by using another resolver)
- Error 16, Censored: it's a third party's choice (the error probably might not be bypassed by using another resolver)
- Error 17, Filtered: it's the client's choice (the error can be bypassed by updating the resolver with client filtering expectations)

Getting more specific than that doesn't help because it wouldn't let a client take action without human intervention. Today, a client might do something like this:
- 15? Repeat the query to a secondary resolver
- 16? Give up
- 17? Log for local admin to review for action ("oh I didn't mean to filter that", "oh my child is going to get grounded for trying to go to that site", etc.)

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 2 above)
Comment (tireddy2) [reply interim-jan-2021/resolver-behavior-self-description.md]: An apt example could be: Cellular provider provides regional specific content (for example TV shows accessible to only users of a specific geographic location).

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 10 above)
Comment (tireddy2) [reply interim-jan-2021/resolver-behavior-self-description.md:18]: > 
> 
> From my perspective, the "designated resolver" upgrade scenario should be completely non-interactive. I think these slides are about supporting other scenarios. 

Okay.

> Some uses I can think of:
> 
>     * Inspecting and debugging my current configuration

Are you referring to OS/IT admin inspecting and debugging the current configuration for troubleshooting ?

> 
>     * Browsing a directory of public resolvers

I think you mean browsing a directory DoH URI services. But browsing requires the end user to iterate through the list of services to make a selection.

Comment (tireddy2) [reply interim-jan-2021/resolver-behavior-self-description.md:18]: > 
> 
> > Several WG members including you opposed any interaction with the end user
> 
> Absolutely correct, and I still feel that way.
> 
> > The argument was an average end user would not understand the privacy terms, server identity and encrypted DNS.
> 
> And this will always continue to be true IMO. Please don't stop your draft direction change :) unless what I'm saying below clarifies my previous "get policy out of my upgrade process" verbiage.
> 
> I agree with Ben: this doesn't mean "make the upgrade process interactive" to me. Rather, this is a way to provide "other info about this resolver intended for humans" to be considered by an auditor or debugger.
> 
> Example: if there was a standard way to return a documentation link, Google DoH servers could return [this page](https://developers.google.com/speed/public-dns/docs/doh/). It would be nice as a client implementor if there was a standard way to retrieve per-resolver documentation.
> 
> Basically, this is about returning man pages that might include terms of server use, not about returning EULAs which require an agreement to the terms of use. The pivot point is about it being a part of the encryption or upgrade negotiation or not. I am definitely opposed to this link being considered a passive agreement (or being made a requirement for active agreement) to terms of use on the server.
> 
> Does this help?

Yes, Thanks. We are in the same page :)

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 13 above)
Comment (tireddy2) [reply interim-jan-2021/resolver-behavior-self-description.md]: Got it.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 15 above)
Comment (tireddy2) [reply interim-jan-2021/resolver-behavior-self-description.md:26]: Agree with both your responses, we proposed https://tools.ietf.org/html/draft-reddy-dnsop-error-page-05 to provide an error page to the user explaining why a DNS query was filtered..

My point is if the DoH server offers multiple DoH services, For example (1) malware blocking and DNSSEC (2) no filtering and no DNSSEC (similar to https://www.quad9.net/doh-quad9-dns-servers/). User has a choice to opt for a) or b) and the user needs to know the above resolver information to make a selection. For instance an home admin could opt for (1) for all the devices attached to the home network.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 14 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md]: Good point. Removing now.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 12 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md]: I'll also remove this line. Other optional behaviors may be uncovered, but we'll cross that bridge when we come to it.

File: interim-jan-2021/resolver-behavior-self-description.md [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+    * Authoritatively if the namespace collides with any global names
+* Ex 1: Enterprise resolvers serving corporation-specific namespaces
+* Ex 2: Public Wi-Fi or cellular networks offering network-local services
+
+## Defining per-namespace optimization
+* Express namespaces for which this resolver provides preferable resolutions
+* Ex 1: ISP routes public content requests to network caches
+* Ex 2: Public resolver designated to serve a namespace to limit parties privy to resolution
+
+## Defining resolver identity
```
Comment (bemasc) [inline interim-jan-2021/resolver-behavior-self-description.md:14]: Speaking of identity, I think this needs to convey the resolver's human-readable name.  I'd also like a favicon :-)

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 3 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md:11]: I mean, feasible at a full granular namespace enumeration or not, a resolver ought to be able to inform clients that it has unique enlightenment for some namespaces for routing to local caches.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 5 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md]: I updated your PR with the non-goal indication; resolving.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 9 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md:12]: Yes, or possibly one that changes the resolution based on party (employees see one portal, everyone else sees a public watered-down version).

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 16 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md:14]: I was about to heckle you about this, but... it would be nice to see something other than an IP address or URI when looking at DNS client configuration UI.

Adding a line.

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 15 above)
Comment (tojens-ietf) [reply interim-jan-2021/resolver-behavior-self-description.md:26]: My view on this is that the templates should be communicated without an attempt to describe their behavior differences in the protocol (since it cannot be automatically be confirmed or reasoned on) and let the user opt-in to a choice via the documentation provided. There should be a default "here's how to upgrade path" that requires no choice.

Basically, info beyond EDE should be considered non-verifiable (given it would require much larger constructs such as a trusted auditing authority) and provided via the documentation for optional exploration by interested users. Today, such behavioral differences are usually provided by different IP addresses (different server identities).

File: interim-jan-2021/resolver-behavior-self-description.md (hunk 3 above)
Comment (bemasc) [reply interim-jan-2021/resolver-behavior-self-description.md:11]: That might be more scalable if it's signaled in the response.  In fact, I believe it's already the case that some "authoritative resolvers" will place their answer in the Authority section (instead of the Answer section) if they were acting as authoritative, not recursive.

I can imagine a draft that says clients can send queries to multiple resolvers, and prefer authoritative answers.  (And perhaps some kind of hint that says "I am authoritative for some zones" without having to enumerate them.)

See also https://tools.ietf.org/html/draft-nottingham-doh-digests-00

Perhaps this is a "goal" but not a "requirement", or a "MAY"-level requirement.

**Comment by bemasc** [comment]
In #37, @jim-add noted that resolver self-description should meet the following description:
* A signalling mechanism to describe resolver (protocol and policy) properties:
   - eg DNSSEC or ECS support, DoH/DoT/DoQ/Do53, split DNS, filtering

I think the proposed requirements already align well with Jim's examples, which is great.
